### PR TITLE
Siridevvscode

### PR DESCRIPTION
**Pull request by sirireddy24**

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -32,7 +32,10 @@
         "onCommand:vscode-sf.deployApplication",
         "onCommand:vscode-sf.publishApplication",
         "onCommand:vscode-sf.removeApplication",
-        "onCommand:vscode-sf.upgradeApplication"
+        "onCommand:vscode-sf.upgradeApplication",
+        "onCommand:vscode-sf.openVSproject",
+        "onCommand:vscode-sf.cleanCsharpApplication",
+        "workspaceContains:*.sfproj"
```
Comment (anantshankar17) [inline package.json]: .sfproj should be defined as a constant at 1 place and used everywhere else.

File: package.json [hunk 2]
```diff
@@ -76,6 +79,16 @@
...
                 "category": "Service Fabric"
+            },
+            {
+                "command": "vscode-sf.openVSproject",
+                "title": "Initialize VS project for VScode",
+                "category": "Service Fabric"
+            },
+            {
+                "command": "vscode-sf.cleanCsharpApplication",
+                "title": "clean csharp Application",
```
Comment (anantshankar17) [inline package.json]: Fix casing. "Clean CSharp Application"

File: package.json [hunk 3]
```diff
@@ -76,6 +79,16 @@
                 "command": "vscode-sf.upgradeApplication",
                 "title": "Upgrade Application",
                 "category": "Service Fabric"
+            },
+            {
+                "command": "vscode-sf.openVSproject",
```
Comment (anantshankar17) [inline package.json:83]: rename vscode-sf.openVSproject to vscode-sf.initializeVSproject

File: src/commands/Initialize VS project for VScode.ts [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+var path = require('path');
+var find = require('find');
+var fs = require('fs');
+import { generatorProject } from '../yo';
+import { getWorkingFolder } from '../yo';
+import { cleanCsharpApplication } from './clean-application-csharp';
+//var parser=new xml2js.Parser();
+export async function openVSproject() {
+  var x = await getWorkingFolder();
+  var files = find.fileSync(/\.sfproj$/, x);
```
Comment (anantshankar17) [inline src/commands/Initialize VS project for VScode.ts]: define "sfproj" at one place.

File: src/commands/Initialize VS project for VScode.ts [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+  if (files.length) {
+  var data = fs.readFileSync(files[0]);
+  var isVS = 1;
+  }
+  else {
+    window.showErrorMessage("This is not a VS project");
```
Comment (anantshankar17) [inline src/commands/Initialize VS project for VScode.ts]: "This project doesn't contain an SFPROJ file."

File: src/commands/Initialize VS project for VScode.ts [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+  var services = [];
+  var serviceProjName = new Array();
+  var servicePackage = new Array();
+  var serviceNameStr = new Array();
+  var csprojdata = new Array();
+  var interfaceprojpath = new Array();
+  var csprojinfo;
+  var appPackagePath = appname;
+  var i;
+  /*var files=await fs.readdirSync(x);
```
Comment (anantshankar17) [inline src/commands/Initialize VS project for VScode.ts]: remove commented lines.

File: src/commands/Initialize VS project for VScode.ts [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+  var appname=files[1];*/
+  var parseString = require('xml2js').parseString;
+  var configjs;
+  await parseString(data, function (err, result) {
+    configjs = result;
+  });
+  var numofservices = Object.keys(configjs["Project"]["ItemGroup"][3]["ProjectReference"]).length;
```
Comment (anantshankar17) [inline src/commands/Initialize VS project for VScode.ts]: What is the "3" that is used here ?

**Comment by sirireddy24** [comment]
3 is the index number of the array “ItemGroup” which has the csproj paths in “ProjectReference”.

Sent from Mail<https://go.microsoft.com/fwlink/?LinkId=550986> for Windows 10

________________________________

File: src/commands/Initialize VS project for VScode.ts [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+    var root = await getWorkingFolder();
+    var p1 = path.join(root, serviceProjName[i], serviceProjName[i] + '.csproj');
+    csprojdata[i] = fs.readFileSync(p1);
+    await parseString(csprojdata[i], function (err, result) {
+      csprojinfo = result;
+    });
+
+    if(csprojinfo["Project"]["PropertyGroup"][0]["TargetFramework"]==undefined)
+    {
+      window.showErrorMessage("Full dotnet projects cannot be opened in VS Code");
```
Comment (anantshankar17) [inline src/commands/Initialize VS project for VScode.ts]: "Only DotNetCore projects can be initialized for VSCode"

File: src/commands/Initialize VS project for VScode.ts [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+    else {
+      services.push({ serviceProjName: serviceProjName[i], servicePackage: servicePackage[i], interfaceprojpath: interfaceprojpath[i] });
+    }
+  };
+  var values = {
+    appname: appname,
+    numofservices: numofservices,
+    services: services,
+  }
+  await fs.writeFileSync(path.join(x, 'vscode-config.json'), JSON.stringify(values), (err) => {
```
Comment (anantshankar17) [inline src/commands/Initialize VS project for VScode.ts]: define at 1 place.

File: src/commands/build-application.ts [hunk 10]
```diff
@@ -97,7 +98,13 @@ function replaceBuildPath(filePath) {
         if (err) {
             return console.log(err);
         }
+        if(!vars._isWindows){
```
Comment (anantshankar17) [inline src/commands/build-application.ts:101]: Why is this required ? This will impact build application for initialize VS / normal Vscode scenarios.

File: src/commands/build-application.ts [hunk 11]
```diff
@@ -134,3 +141,23 @@ async function createPublishProfile() {
         console.log('Completed!');
     });
 }
+
+
+/*function dumpfolder(){
```
Comment (anantshankar17) [inline src/commands/build-application.ts]: remove dead code.

File: src/extension.ts [hunk 12]
```diff
@@ -9,15 +10,30 @@ import { deployApplication } from './commands/deploy-application';
...
+var find=require('find');
+var path=require('path');
+import {getWorkingFolder} from './yo';
 // This method is called when your extension is activated. Activation is
 // controlled by the activation events defined in package.json.
-export function activate(context: ExtensionContext) {
-
+export async function activate(context: ExtensionContext) {
+    var root=await getWorkingFolder();
+    var files = find.fileSync(/\.sfproj$/,root);
```
Comment (anantshankar17) [inline src/extension.ts]: Use constant.

File: src/extension.ts [hunk 13]
```diff
@@ -9,15 +10,30 @@ import { deployApplication } from './commands/deploy-application';
...
     // Use the console to output diagnostic information (console.log) and errors (console.error).
     // This line of code will only be executed once when your extension is activated.
     console.log('Congratulations, your extension "Service Fabric" is now active!');
-
+    if(files.length&&!fs.existsSync(path.join(root,'vscode-config.json'))){
```
Comment (anantshankar17) [inline src/extension.ts]: Use predefined constant.

**Comment by sirireddy24** [comment]
The xcopy command which we are appending works only for back slash in case of windows, so we shouldn’t replace it with a forward slash if it is windows.

Sent from Mail<https://go.microsoft.com/fwlink/?LinkId=550986> for Windows 10

________________________________

File: src/extension.ts [hunk 14]
```diff
@@ -9,15 +10,30 @@ import { deployApplication } from './commands/deploy-application';
...
+        window.showInformationMessage("This is a VS project, do you want to initialise it for VS code?","Continue","Cancel").then(choice => {
+            if (choice === 'Continue') {
+                openVSproject().then(res => cleanCsharpApplication());
```
Comment (anantshankar17) [inline src/extension.ts]: You have already called clean earlier. remove redundant calls.

File: src/yo/index.ts [hunk 15]
```diff
@@ -116,7 +142,8 @@ function list(yo: Yeoman): Promise<QuickPickItem[]> {
 					return {
 						label: generator.name.replace(/(^|\/)generator\-/i, '$1') as string,
 						description: generator.description,
-						subGenerators: generator.subGenerators
+						subGenerators: generator.subGenerators,
```
Comment (anantshankar17) [inline src/yo/index.ts:143]: remove unnescessary updates.

File: src/commands/Initialize VS project for VScode.ts [hunk 16]
```diff
@@ -0,0 +1,107 @@
+import * as vscode from 'vscode';
```
Comment (anantshankar17) [inline src/commands/Initialize VS project for VScode.ts]: Rename the file to InitializeVSProject.ts

File: src/commands/Initialize VS project for VScode.ts [hunk 17]
```diff
@@ -0,0 +1,107 @@
...
+  var configjs;
+  await parseString(data, function (err, result) {
+    configjs = result;
+  });
+  var l = 0;
+  for (l = 0; l < configjs["Project"]["ItemGroup"].length; l++) {
+    if (configjs["Project"]["ItemGroup"][l]["ProjectReference"] != undefined)
+      break;
+  }
+  console.log(l);
```
Comment (anantshankar17) [inline src/commands/Initialize VS project for VScode.ts]: remove stray log cmds.

File: src/commands/Initialize VS project for VScode.ts [hunk 18]
```diff
@@ -0,0 +1,107 @@
...
+  var numofservices = Object.keys(configjs["Project"]["ItemGroup"][l]["ProjectReference"]).length;
+  for (i = 0; i < numofservices; i++) {
+    serviceNameStr[i] = configjs["Project"]["ItemGroup"][l]["ProjectReference"][i]["$"]["Include"];
+  }
+  for (i = 0; i < numofservices; i++) {
```
Comment (anantshankar17) [inline src/commands/Initialize VS project for VScode.ts]: move the code of reading the sfproj and populating the data structure to a new method in this file. Call that method from here.

File: src/yo/index.ts [hunk 19]
```diff
@@ -23,18 +23,32 @@ async function getWorkingFolder() {
 	return selectedWkFolder ? selectedWkFolder.uri.fspath : undefined;
 }
 
-export async function generatorProject(addService) {
+
+export async function generatorProject(addService,openVSproject) {
 	const cwd = await getWorkingFolder();
+	const configfilename='vscode-config.json';
```
Comment (anantshankar17) [inline src/yo/index.ts]: refer the constant defined at 1 place.

File: src/commands/Initialize VS project for VScode.ts (hunk 18 above)
Comment (anantshankar17) [reply src/commands/Initialize VS project for VScode.ts]: Leave, this for now. Complete other comments first.

File: package.json [hunk 20]
```diff
@@ -32,7 +32,10 @@
         "onCommand:vscode-sf.deployApplication",
         "onCommand:vscode-sf.publishApplication",
         "onCommand:vscode-sf.removeApplication",
-        "onCommand:vscode-sf.upgradeApplication"
+        "onCommand:vscode-sf.upgradeApplication",
+        "onCommand:vscode-sf.openVSproject",
+        "onCommand:vscode-sf.cleanCsharpApplication",
```
Comment (anantshankar17) [inline package.json]: remove cleanCsharpApplication cmd.
